proposals/p1885.md

```diff
@@ -0,0 +1,458 @@
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/1885)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+    -   [Other languages](#other-languages)
+        -   [C++](#c)
+        -   [Python](#python)
+        -   [Rust](#rust)
+        -   [Typescript](#typescript)
+        -   [Go](#go)
+-   [Proposal](#proposal)
+-   [Details](#details)
+    -   [Indexable](#indexable)
+        -   [Range-based `for` behavior](#range-based-for-behavior)
+        -   [Interface and usage](#interface-and-usage)
+        -   [Implicit `[]` support](#implicit--support)
+        -   [Mutable access](#mutable-access)
+    -   [Consumable](#consumable)
+        -   [Range-based `for` behavior](#range-based-for-behavior-1)
+        -   [Interface and usage](#interface-and-usage-1)
+        -   [TODO](#todo)
+    -   [Iterable](#iterable)
+        -   [Range-based `for` behavior](#range-based-for-behavior-2)
+        -   [Interface and usage](#interface-and-usage-2)
+        -   [Iterator: `begin()` & `end()`](#iterator-begin--end)
+        -   [Mutable, Bidirectional, and others](#mutable-bidirectional-and-others)
+        -   [Chaining & Lazy evaluation](#chaining--lazy-evaluation)
+    -   [Limitations and future language improvements](#limitations-and-future-language-improvements)
+    -   [TODO](#todo-1)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [Using interface type parameters](#using-interface-type-parameters)
+    -   [Other](#other)
+
+<!-- tocstop -->
+
+## Problem
+
+The current `for` proposal does not define a way for user-defined types to
+support range-based for loops. Adding support would make it significantly easier
+to define custom containers, as well as containers in Carbon standard library.
+
+The goals for the solution includes:
+
+-   Expressive and easy to implement
+-   Minimal assumptions and requirements on the user-defined type
+-   Minimal impact to maintain and use the type after (loose coupling, minimal
+    impact on interface, type content, ...)
+-   As well as the standard: Easy to use right, hard to use incorrectly,
+    performance, don't pay for what you don't want, ...
+
+While keeping in mind for the future:
+
+-   Compatible with future algorithms (`sort`, `filter`, `transform`, ...)
+-   Suitable for chaining (`std::ranges::views`)
+-   Suitable for lazy evaluation / generators
+
+## Background
+
+The original proposal of `for` loops were discussed in
+[p0353](/proposals/p0353.md), and the basic design documented in
+[control_flow/loops#for](/docs/design/control_flow/loops.md#for).
+
+The resulting syntax is:
+
+> `for (` _var declaration_ `in` _expression_ `) {` _statements_ `}`
+
+The original `for` proposal does not define a way for user-defined types to
+opt-in to support range-based for loops. An approach is proposed in this
+document.
+
```

Comment:
Added a small paragraph to that effect. Do you think this challenges the iterator approach in this proposal? I can imagine a situation where this could work provided the "container" implements a specific 'indexable' interface (removed from here, and a WIP by @geoffromer IIRC), that would likely put an unecessary random-access constraint on the said user-container.

Reply:
I just realized one major advantage of the "cursor" model over the "iterator" model: it's really hard to see how array r-values could have iterators, since r-values don't have addresses, and an iterator is the functional equivalent of a pointer.

Of course, this problem goes beyond `for` loops: it implies we can't run STL algorithms over array r-values, so we might need to solve this some other way anyway. But I think the cursor model needs to at least be in "alternatives considered", if not the primary proposal.

Reply:
Switched to cursor, and iterator now in alternatives. Mentioned R-value simplification.